src/core/connection.c

```diff
@@ -1892,6 +1892,223 @@ QuicConnRestart(
+QuicConnSendResumptionTicket(
+    _In_ QUIC_CONNECTION* Connection,
+    _In_ uint16_t AppDataLength,
+    _In_reads_bytes_opt_(AppDataLength)
+        const uint8_t* AppResumptionData
+    )
+{
+    QUIC_STATUS Status;
+    uint32_t EncodedTransportParametersLength = 0;
+    uint8_t* TicketBuffer = NULL;
+
+    QUIC_TLS_RESUMPTION_TICKET Ticket = { 0, };
+    Ticket.TicketVersion = QUIC_TLS_RESUMPTION_TICKET_VERSION;
+    Ticket.QuicVersion = QUIC_VERSION_LATEST;
+    Ticket.AlpnLength = *(Connection->Crypto.TlsState.NegotiatedAlpn);
+    Ticket.AppTicketLength = AppDataLength;
+
+    QUIC_TRANSPORT_PARAMETERS LocalTPCopy;
+    QuicCopyMemory(&LocalTPCopy, &Connection->LocalTP, sizeof(LocalTPCopy));
+    LocalTPCopy.Flags = LocalTPCopy.Flags & (
+        QUIC_TP_FLAG_ACTIVE_CONNECTION_ID_LIMIT |
+        QUIC_TP_FLAG_INITIAL_MAX_DATA |
+        QUIC_TP_FLAG_INITIAL_MAX_STRM_DATA_BIDI_LOCAL |
+        QUIC_TP_FLAG_INITIAL_MAX_STRM_DATA_BIDI_REMOTE |
+        QUIC_TP_FLAG_INITIAL_MAX_STRM_DATA_UNI |
+        QUIC_TP_FLAG_INITIAL_MAX_STRMS_BIDI |
+        QUIC_TP_FLAG_INITIAL_MAX_STRMS_UNI);
+
+    const uint8_t* EncodedLocalTP = QuicCryptoTlsEncodeTransportParameters(
+        Connection,
+        &LocalTPCopy,
+        &EncodedTransportParametersLength);
+
+    if (EncodedLocalTP == NULL) {
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
+        goto Error;
+    }
+    Ticket.TransportParamsLength = EncodedTransportParametersLength;
+
+    uint32_t TotalTicketLength = (uint32_t) (QuicVarIntSize(Ticket.TicketVersion) +
+        QuicVarIntSize(Ticket.AlpnLength) +
+        QuicVarIntSize(Ticket.TransportParamsLength) +
+        QuicVarIntSize(Ticket.AppTicketLength) +
+        sizeof(Ticket.QuicVersion) +
+        Ticket.AlpnLength +
+        Ticket.TransportParamsLength +
+        Ticket.AppTicketLength);
+
+    TicketBuffer = QUIC_ALLOC_NONPAGED(TotalTicketLength);
+    if (TicketBuffer == NULL) {
+        QuicTraceEvent(
+            AllocFailure,
+            "Allocation of '%s' failed. (%llu bytes)",
+            "Server resumption ticket",
+            TotalTicketLength);
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
+        goto Error;
+    }
+
+    /*
+        Encoded ticket format is as follows:
+        Ticket Version (QUIC_VAR_INT) [1..4]
+        Quic Version [4]
+        Negotiated ALPN length (QUIC_VAR_INT) [1..4]
+        Negotiated ALPN [...]
+        Transport Parameters length (QUIC_VAR_INT) [1..4]
+        Transport Parameters [...]
+        App Ticket length (QUIC_VAR_INT) [1..4]
+        App Ticket (omitted if length is zero) [...]
+    */
+    uint8_t* TicketCursor = QuicVarIntEncode(Ticket.TicketVersion, TicketBuffer);
+    *(uint32_t*)TicketCursor = QuicByteSwapUint32(Ticket.QuicVersion);
+    TicketCursor += sizeof(Ticket.QuicVersion);
+    TicketCursor = QuicVarIntEncode(Ticket.AlpnLength, TicketCursor);
+    QuicCopyMemory(TicketCursor, Connection->Crypto.TlsState.NegotiatedAlpn + 1, (size_t) Ticket.AlpnLength);
```

Comment:
Unfortunately, on x86 and arm builds, this generates a warning without the cast because size_t on those platforms is 4 bytes instead of 8 bytes. If you have a better solution, I'm open to it.
I also noticed that QUIC_VAR_INT isn't defined in quic_tls.h so I had to use `uint64_t` for these fields.